consensus/src/metrics_safety_rules.rs

```diff
@@ -34,17 +34,32 @@ impl MetricsSafetyRules {
 
     pub fn perform_initialize(&mut self) -> Result<(), Error> {
         let consensus_state = self.consensus_state()?;
-        let sr_waypoint = consensus_state.waypoint();
-        let proofs = self
-            .storage
-            .retrieve_epoch_change_proof(sr_waypoint.version())
-            .map_err(|e| {
-                Error::InternalError(format!(
-                    "Unable to retrieve Waypoint state from storage, encountered Error:{}",
-                    e
-                ))
-            })?;
-        self.initialize(&proofs)
+        let mut waypoint_version = consensus_state.waypoint().version();
+        loop {
+            let proofs = self
+                .storage
+                .retrieve_epoch_change_proof(waypoint_version)
+                .map_err(|e| {
+                    Error::InternalError(format!(
+                        "Unable to retrieve Waypoint state from storage, encountered Error:{}",
+                        e
+                    ))
+                })?;
+            // We keep initializing safety rules as long as the waypoint continues to increase.
+            // This is due to limits in the number of epoch change proofs that storage can provide.
+            match self.initialize(&proofs) {
+                Err(Error::WaypointOutOfDate(
+                    prev_version,
+                    curr_version,
+                    _, /* current epoch */
```

Comment:
doesn't seem we're using these two fields?

Reply:
@zekun000 You meant `current_epoch` and  `provided_epoch`? Yeah we're not using them at the moment. I tried to put/expose more info in the error return in case we may need them in future (e.g., debugging, logging, etc.). If you think we should remove them for now, I can do so!

Reply:
I think we can add some logging here like "SafetyRules epoch is updated from {} to {}"

Reply:
having some logs per run could give some hints when something goes wrong (e.g. infinite loop). it happened before when we're stuck in an infinite loop and had no idea where that loop is

Reply:
@zekun000 I'm going to add logging in this PR. Will update the PR soon.

Reply:
Added log to record waypoint version and epoch info.